### ISSUE TITLE
Demo Colab/Notebook

### ISSUE THREAD

**Issue by asigalov61** [Natooz/MidiTok#6]
Hey Nathan,

So I wanted to contribute by making a little full workflow colab for MidiTok.

Here is a first draft. Let me know what you think, please.

Also, I am experiencing issues saving midis when using the tokens_to_midi function. See last lines in colab

https://github.com/Tegridy-Code/MidiTok/blob/main/colab-notebooks/MidiTok_Full_Workflow_Tutorial.ipynb

Alex.

**Comment by Natooz** [comment]
Hi Alex,

Colab notebooks are gratefully welcomed !
When yours is done just send a PR !

About your issues when saving MIDIs (tokens_to_midi), can you give more details on that ?
I'm guessing it might be because of "invalide" token types successions (eg at some point the model predicted a Pitch token then a Position instead of Velocity). In that case MidiTok ignores the bad type successions.
[Here is the REMI tokens_to_midi method](https://github.com/Natooz/MidiTok/blob/main/miditok/remi.py#L128) to give some hints (see line 168 the try bloc).

**Comment by asigalov61** [comment]
Hey Nathan,

Hmm... not sure what happened but I can't replicate the save issue anymore so I am going to assume that it was something on my end and everything works properly atm. My bad. Maybe it was some temp code conflict or something on my end...

I will PR the finished v. 1.0 colab shortly. See if you like it. It works pretty well.

Alex
________________________________

**Comment by envilk** [comment]
First of all, thanks @Natooz for Miditok implementation of the Pop Music Transformer, it was the only way I could use this model. Also thanks to @asigalov61 for developing an example collab notebook free to modify and apply to your data.

I have three questions regarding a snippet of the notebook code I'm struggling to understand:

```python
train_data = []

train_data.extend(tokens[0])

val_dataset = train_data[:int(len(train_data) * 0.5)]
test_dataset = train_data[:int(len(train_data) * 0.5)]

train_list = train_data
val_list = val_dataset
test_list = []
print('=' * 50)

print('Processing INTs datasets...')
train_dataset = EPianoDataset(train_list, max_seq, random_seq)
val_dataset = EPianoDataset(val_list, max_seq)
test_dataset = EPianoDataset(test_list, max_seq)
print('=' * 50)

print('Loading INTs datasets...')
batch_size = number_of_batches
train_loader = DataLoader(train_dataset, batch_size=batch_size, num_workers=n_workers, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=batch_size, num_workers=n_workers)
test_loader = DataLoader(test_dataset, batch_size=batch_size, num_workers=n_workers)
```

1) Why ```val_dataset``` and ```test_dataset``` are the same training data split?
2) Why ```test_dataset``` is filled up with a training data split at the beginning, and then an empty list is assigned to it in the ```EPianoDataset``` constructor line?
3) Why ```test_dataset``` is not used in the rest of the notebook?

Thanks in advance!

**Comment by envilk** [comment]
Thanks for the answer @Natooz!

Now I understand more about Miditok, I will try splitting the dataset as you have told me!

**Comment by Natooz** [comment]
You're welcome !

As Alex didn't answered I'll add that it's common practice to use a valid subset to monitor a model's performance during training (via the metrics you want: loss, accuracy, perplexity, F1 score ...) to avoid overfitting. The samples from the valid metrics aren't used to train the model (meaning adapt its parameters) so one should want it to perform well on unseen data.
The test subset is then used after training to fully test / assess a model, on defined metrics again.
